File: Data Structures & Algorithms/search-for-word/submission-4.js
class Solution {
    /**
     * @param {character[][]} board
     * @param {string} word
     * @return {boolean}
     */
    exist(board, word) {
        const ROWS = board.length
        const COLS = board[0].length
        const path = new Set();
        const dfs = (r,c,i) => {
            if(i === word.length){
                return true;
            }
            if(r >= ROWS || c >= COLS || 0 > r || 0 > c 
                || path.has(`${r},${c}`) || board[r][c] !== word[i]){
                    return false
                }
            path.add(`${r},${c}`)
            const ret = dfs(r + 1, c, i + 1) || 
                        dfs(r, c + 1, i + 1) ||
                        dfs(r - 1, c, i + 1) ||
                        dfs(r, c - 1, i + 1)
            path.delete(`${r},${c}`)
            return ret
        }
        for(let i = 0; i < ROWS; i++){
            for(let j = 0; j < COLS; j++){
                if(board[i][j] == word[0] && dfs(i,j,0)){
                    return true
                }
            }
        }
        return false
    }
}

File: Data Structures & Algorithms/surrounded-regions/submission-4.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        ROWS = len(board)
        COLS = len(board[0])
        visit = set()
        def dfs(r,c):
            if(r < 0 or r >= ROWS or c < 0 or c >= COLS or (r,c) in visit or 
                board[r][c] == "X"):
                return
            visit.add((r,c))
            board[r][c] = "F"
            dfs(r+1,c)
            dfs(r-1,c)
            dfs(r,c+1)
            dfs(r,c-1)

        for j in range(COLS):
            for i in [0, ROWS - 1]:
                if(board[i][j] == "O"):
                    dfs(i,j)
        
        for j in range(ROWS):
            for i in [0, COLS - 1]:
                if(board[j][i] == "O"):
                    dfs(j,i)
        
        for i in range(ROWS):
            for j in range(COLS):
                if(board[i][j] == "O"):
                    board[i][j] = "X"
                elif(board[i][j] == "F"):
                    board[i][j] = "O"

        

